crates/node/src/cluster_state_refresher/mod.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{collections::BTreeMap, sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use tokio::{
+    sync::watch,
+    time::{self, Instant, Interval, MissedTickBehavior},
+};
+use tracing::{debug, instrument, trace};
+
+use restate_core::{
+    cancellation_watcher,
+    network::{
+        Incoming, MessageRouterBuilder, MessageStream, NetworkSender, Networking, Outgoing,
+        TransportConnect,
+    },
+    Metadata, ShutdownError, TaskCenter, TaskKind,
+};
+use restate_types::{
+    cluster::cluster_state::{AliveNode, ClusterState, DeadNode, NodeState, SuspectNode},
+    config::{CommonOptions, Configuration},
+    net::cluster_state::{NodePing, NodePong},
+    time::MillisSinceEpoch,
+    GenerationalNodeId, PlainNodeId,
+};
+
+//todo(azmy): make these configurable
+const SUSPECT_THRESHOLD_FACTOR: u32 = 2;
+const DEAD_THRESHOLD_FACTOR: u32 = 4;
+
+#[derive(Clone)]
+struct SeenState {
+    generational_node_id: GenerationalNodeId,
+    seen_at: MillisSinceEpoch,
+}
+
+impl SeenState {
+    fn new(generational_node_id: GenerationalNodeId) -> Self {
+        SeenState {
+            generational_node_id,
+            seen_at: MillisSinceEpoch::now(),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+struct NodeTracker {
+    seen: Option<SeenState>,
+    last_attempt_at: Option<MillisSinceEpoch>,
+}
+
+pub struct ClusterStateRefresher<T> {
+    metadata: Metadata,
+    ping_requests: MessageStream<NodePing>,
+    ping_responses: MessageStream<NodePong>,
+    networking: Networking<T>,
+    nodes: BTreeMap<PlainNodeId, NodeTracker>,
+    heartbeat_interval: Duration,
+    cluster_state_watch_tx: watch::Sender<Arc<ClusterState>>,
+    start_time: MillisSinceEpoch,
+}
+
+impl<T> ClusterStateRefresher<T>
+where
+    T: TransportConnect,
+{
+    pub(crate) fn new(
+        metadata: Metadata,
+        networking: Networking<T>,
+        router_builder: &mut MessageRouterBuilder,
+    ) -> Self {
+        let config = Configuration::pinned();
+        ClusterStateRefresher {
+            metadata,
+            ping_requests: router_builder.subscribe_to_stream(128),
+            ping_responses: router_builder.subscribe_to_stream(128),
+            networking,
+            nodes: BTreeMap::default(),
+            heartbeat_interval: config.common.heartbeat_interval.into(),
+            cluster_state_watch_tx: watch::Sender::new(Arc::new(ClusterState::empty())),
```

Comment:
Is there a reason why ClusterState is in an Arc?

Reply:
Borrowed values from a watch has a Read lock which means you should only borrow for very short period of times and never across await points. Hence I think using an Arc is safer so we can cheaply copy the cluster state and release the borrow. Then you can pass this state snapshot around or use it across await points